src/test/java/com/make/my/day/hm5/MergeSortJoinTest.java

```diff
@@ -0,0 +1,226 @@
+    Integer flag = 0;
+    L nextLeft;
+    R nextRight;
+    L prevLeft;
+    boolean last = true;
+
+    public MergeSortInnerJoinSpliterator(Stream<L> left,
+                                         Stream<R> right,
+                                         Function<L, C> keyExtractorLeft,
+                                         Function<R, C> keyExtractorRight,
+                                         boolean isSorted) {
+
+      if (!isSorted) {
+        this.leftIterator = left.sorted()
+                                .iterator();
+        this.rightIterator = right.sorted()
+                                  .iterator();
+      } else {
+        this.leftIterator = left.iterator();
+        this.rightIterator = right.iterator();
+      }
+      this.keyExtractorLeft = keyExtractorLeft;
+      this.keyExtractorRight = keyExtractorRight;
+
+      if (leftIterator.hasNext()) {
+        nextLeft = leftIterator.next();
+      }
+
+    }
+
+    @Override
+    public boolean tryAdvance(Consumer<? super Pair<L, R>> action) {
+      if (leftIterator.hasNext() || rightIterator.hasNext() || last) {
+        if (flag == 0) {
+          if (leftIterator.hasNext() && rightIterator.hasNext()) {
+            prevLeft = nextLeft;
+            nextLeft = leftIterator.next();
+            nextRight = rightIterator.next();
+          }
+        } else if (flag == -1) {
+          if (leftIterator.hasNext()) {
+            prevLeft = nextLeft;
+            nextLeft = leftIterator.next();
+          } else if (last) {
+            prevLeft = nextLeft;
+            nextLeft = null;
+            last = false;
+          } else {
+            return false;
+          }
+        } else {
+          if (rightIterator.hasNext()) {
+            nextRight = rightIterator.next();
+          } else {
+            return false;
+          }
+        }
+
+        if (isLeftEqualsRight()) {
+          if (nextLeft != null && keyExtractorLeft.apply(prevLeft)
+                                                  .equals(keyExtractorLeft.apply(nextLeft))) {
+            action.accept(new Pair<>(prevLeft, nextRight));
+          }
+          action.accept(new Pair<>(prevLeft, nextRight));
+          flag = 1;
+          return true;
+        } else {
+          return true;
+        }
+      }
+
+      return false;
+    }
+
+    private boolean isLeftEqualsRight() {
+      int compare = keyExtractorLeft.apply(prevLeft)
+                                    .compareTo(keyExtractorRight.apply(nextRight));
+      if (compare < 0) {
+        flag = -1;
+        return false;
+      }
+      if (compare > 0) {
+        flag = 1;
+        return false;
+      }
+      return true;
+    }
+
+    @Override
+    public Spliterator<Pair<L, R>> trySplit() {
+      return null;
+    }
+
+    @Override
+    public long estimateSize() {
+      return 0;
+    }
+
+    @Override
+    public int characteristics() {
+      return SIZED | SUBSIZED | CONCURRENT | IMMUTABLE | ORDERED;
```

Comment:
Why is it CUNCURRENT?

Reply:
it's just a bitmask, it does not have to be concurrent